test/stacktraces/stacktraces.test.js

```diff
@@ -0,0 +1,343 @@
+//    force this by building the sourcemap with a bogus "no-such-dir"
+//    "sourceRoot" (see the setting in fixtures/tsconfig.json).
+//
+// Note: Other sourcemap tests are in "test/sourcemaps/". That set currently
+// lacks a test that "sourcesContent" works -- which is one of the things tested
+// by this case.
+tape.test('error.exception.stacktrace with sourcemap', function (t) {
+  const server = new MockAPMServer()
+  server.start(function (serverUrl) {
+    execFile(
+      process.execPath,
+      ['fixtures/dist/throw-an-error-with-sourcemap.js'],
+      {
+        cwd: __dirname,
+        timeout: 3000,
+        env: Object.assign({}, process.env, {
+          ELASTIC_APM_SERVER_URL: serverUrl
+        })
+      },
+      function done (err, _stdout, _stderr) {
+        t.ok(err, 'throw-an-error-with-sourcemap.js errored out')
+        t.ok(/Error: boom/.test(err.message),
+          'err.message includes /Error: boom/: ' + JSON.stringify(err.message))
+        t.ok(server.events[0].metadata, 'APM server got event metadata object')
+        t.ok(server.events[1].error, 'APM server got error event')
+        const stacktrace = server.events[1].error.exception.stacktrace
+        t.deepEqual(
+          stacktrace[0],
+          {
+            filename: path.join('fixtures', 'dist', 'no-such-dir', 'throw-an-error-with-sourcemap.ts'),
+            lineno: 16,
+            function: 'main',
+            library_frame: false,
+            abs_path: path.join(__dirname, 'fixtures', 'dist', 'no-such-dir', 'throw-an-error-with-sourcemap.ts'),
+            pre_context: ['', 'function main(msg: string) {'],
+            context_line: '  throw new Error(msg)',
+            post_context: ['}', '']
+          },
+          'stacktrace top frame is as expected'
+        )
+        server.close()
+        t.end()
+      }
+    )
+  })
+})
+
+tape.test('stackTraceFromErrStackString()', function (t) {
+  function theFunction () {
+    return new Error('here I am') // <-- the top frame will point here
+  }
+
+  // To avoid the "lineno" test below failing frequently whenever this file is
+  // edited, we read this file to get the current line number of the
+  // `new Error('...')` line above.
+  const lines = fs.readFileSync(__filename, { encoding: 'utf8' }).split(/\n/g)
+  let lineno = null
+  for (let i = 0; i < lines.length; i++) {
+    if (/^\s*return new Error\('here I am'\)/.test(lines[i])) {
+      lineno = i + 1
+      break
+    }
+  }
+
+  const stacktrace = stackTraceFromErrStackString(log, theFunction())
+  t.ok(stacktrace, 'got a stacktrace')
+  t.ok(Array.isArray(stacktrace), 'stacktrace is an Array')
+  t.deepEqual(
+    stacktrace[0],
+    {
+      filename: path.relative(process.cwd(), __filename),
+      function: 'theFunction',
+      lineno: lineno,
+      library_frame: false,
+      abs_path: __filename
+    },
+    'stacktrace top frame is as expected'
+  )
+
+  t.end()
+})
+
+tape.test('gatherStackTrace()', function (suite) {
```

Comment:
REVIEW-NOTE This test case replaces '#parseCallsite()' test from test/parsers.js